<h1 style="font-family: Trebuchet MS; padding: 12px; font-size: 30px; color: #370617; text-align: center; line-height: 1.25;">Recommender System<br><span style="color: #E50914; font-size: 60px"><b>NETFLIX Recommendation</b></span><br><span style="color: #03071e; font-size: 20px">⚫️🔴Movie and TV Show🔴⚫️</span></h1>

<h1 style="font-family: Trebuchet MS; font-size: 14px; color: #370617; text-align: right; ">By: Muhammad Faarisul Ilmi</h1>
<hr>

<h1 style="font-family: Trebuchet MS; font-size: 20px; color: #e85d04; text-align: center; "><b>If you find this notebook useful, give it a thumbs up 😉👍🏻</b></h1>

<p style="background-color:#E50914;font-family:Trebuchet MS;font-weight:bold;color:#ffffff;font-size:40px;text-align:center;border-radius:100px 100px">Table of Contents</p>

In this notebook, we will cover:
* [Overview](#0)
* [Importing Libraries](#1)
* [Loading Dataset](#2)
* [Recommender System](#3)
* [Result](#4)

<a id="0"></a>
<p style="background-color:#E50914;font-family:Trebuchet MS;font-weight:bold;color:#ffffff;font-size:40px;text-align:center;border-radius:100px 100px">Overview</p>

![img](https://thephrase.id/wp-content/uploads/2022/03/7dc497e2-4975-11ec-a9ce-066b49664af6_cm_1440w.jpg)
In this notebook, we are going to build a recommender system based on [Netflix](https://www.kaggle.com/datasets/shivamb/netflix-shows) dataset. There are several types of recommender systems, one of which is Content-Based Filtering. That's what we focus on in this particular notebook since we are not dealing with any user data such as user's rating and review. The idea is to get the features of each item (content) and give the user a recommendation based on the similarity between them

**Let's jump right into the code**




![](https://www.researchgate.net/profile/Lionel-Ngoupeyou-Tondji/publication/323726564/figure/fig5/AS:631605009846299@1527597777415/Content-based-filtering-vs-Collaborative-filtering-Source.png)

<a id="1"></a>
<p style="background-color:#E50914;font-family:Trebuchet MS;font-weight:bold;color:#ffffff;font-size:40px;text-align:center;border-radius:100px 100px">Importing Libraries</p>

In [1]:
import string
import numpy as np
import pandas as pd
import sklearn

import warnings
warnings.filterwarnings('ignore')

<a id="2"></a>
<p style="background-color:#E50914;font-family:Trebuchet MS;font-weight:bold;color:#ffffff;font-size:40px;text-align:center;border-radius:100px 100px">Loading Dataset</p>

In [2]:
df = pd.read_csv('../input/netflix-shows/netflix_titles.csv')
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [3]:
df.describe(include='all')

show_id   type                 title       director  \
count     8807   8807                  8807           6173   
unique    8807      2                  8807           4528   
top         s1  Movie  Dick Johnson Is Dead  Rajiv Chilaka   
freq         1   6131                     1             19   
mean       NaN    NaN                   NaN            NaN   
std        NaN    NaN                   NaN            NaN   
min        NaN    NaN                   NaN            NaN   
25%        NaN    NaN                   NaN            NaN   
50%        NaN    NaN                   NaN            NaN   
75%        NaN    NaN                   NaN            NaN   
max        NaN    NaN                   NaN            NaN   

                      cast        country       date_added  release_year  \
count                 7982           7976             8797   8807.000000   
unique                7692            748             1767           NaN   
top     David Attenborough  United States  January 1, 2020           NaN   
freq                    19           2818              109           NaN   
mean                   NaN            NaN              NaN   2014.180198   
std                    NaN            NaN              NaN      8.819312   
min                    NaN            NaN              NaN   1925.000000   
25%                    NaN            NaN              NaN   2013.000000   
50%                    NaN            NaN              NaN   2017.000000   
75%                    NaN            NaN              NaN   2019.000000   
max                    NaN            NaN              NaN   2021.000000   

       rating  duration                     listed_in  \
count    8803      8804                          8807   
unique     17       220                           514   
top     TV-MA  1 Season  Dramas, International Movies   
freq     3207      1793                           362   
mean      NaN       NaN                           NaN   
std       NaN       NaN                           NaN   
min       NaN       NaN                           NaN   
25%       NaN       NaN                           NaN   
50%       NaN       NaN                           NaN   
75%       NaN       NaN                           NaN   
max       NaN       NaN                           NaN   

                                              description  
count                                                8807  
unique                                               8775  
top     Paranormal activity at a lush, abandoned prope...  
freq                                                    4  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [5]:
pd.DataFrame({'Total missing values':df.isna().sum(),
              'Percentage':(df.isna().sum()/len(df))*100})

Total missing values  Percentage
show_id                          0    0.000000
type                             0    0.000000
title                            0    0.000000
director                      2634   29.908028
cast                           825    9.367549
country                        831    9.435676
date_added                      10    0.113546
release_year                     0    0.000000
rating                           4    0.045418
duration                         3    0.034064
listed_in                        0    0.000000
description                      0    0.000000

<a id="3"></a>
<p style="background-color:#E50914;font-family:Trebuchet MS;font-weight:bold;color:#ffffff;font-size:40px;text-align:center;border-radius:100px 100px">Recommender System</p>

<h1 style="font-family: Trebuchet MS; font-size: 25px; color: #a4161a; text-align: left; "><b>Feature Selection</b></h1>

We won't use all the columns or features for this notebook. So then, the recommendation that we give to the users will only consider the information contained in the following columns:
- Type
- Director
- Rating
- Listed_in
- Description

In [6]:
new_df = df[['title', 'type', 'director', 'cast', 'rating', 'listed_in', 'description']]
new_df.set_index('title', inplace=True)
new_df.head()

type         director  \
title                                             
Dick Johnson Is Dead     Movie  Kirsten Johnson   
Blood & Water          TV Show              NaN   
Ganglands              TV Show  Julien Leclercq   
Jailbirds New Orleans  TV Show              NaN   
Kota Factory           TV Show              NaN   

                                                                    cast  \
title                                                                      
Dick Johnson Is Dead                                                 NaN   
Blood & Water          Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
Ganglands              Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
Jailbirds New Orleans                                                NaN   
Kota Factory           Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   

                      rating  \
title                          
Dick Johnson Is Dead   PG-13   
Blood & Water          TV-MA   
Ganglands              TV-MA   
Jailbirds New Orleans  TV-MA   
Kota Factory           TV-MA   

                                                               listed_in  \
title                                                                      
Dick Johnson Is Dead                                       Documentaries   
Blood & Water            International TV Shows, TV Dramas, TV Mysteries   
Ganglands              Crime TV Shows, International TV Shows, TV Act...   
Jailbirds New Orleans                             Docuseries, Reality TV   
Kota Factory           International TV Shows, Romantic TV Shows, TV ...   

                                                             description  
title                                                                     
Dick Johnson Is Dead   As her father nears the end of his life, filmm...  
Blood & Water          After crossing paths at a party, a Cape Town t...  
Ganglands              To protect his family from a powerful drug lor...  
Jailbirds New Orleans  Feuds, flirtations and toilet talk go down amo...  
Kota Factory           In a city of coaching centers known to train I...

If you take a look at the missing values in this dataset, you will realize that the director column has 2634 NaN values which correspond with almost 30 percents of total data in that column. So, we can't just drop the NaN values because we will lose lots of movies to be given, instead we just fill the NaN values with empty string

In [7]:
new_df.fillna('', inplace=True)

In [8]:
# For director, cast, and listed_in
# Because there is more than 1 people and categories
# We don't want if people share the same first or last name consider the same person
# or the word that appear in many categories (TV, etc) consider the same category
def separate(texts):
    t = []
    for text in texts.split(','):
        t.append(text.replace(' ', '').lower())
    return ' '.join(t)

def remove_space(texts):
    return texts.replace(' ', '').lower()

def remove_punc(texts):
    return texts.translate(str.maketrans('','',string.punctuation)).lower()

In [9]:
new_df['type'] = new_df['type'].apply(remove_space)
new_df['director'] = new_df['director'].apply(separate)
new_df['cast'] = new_df['cast'].apply(separate)
new_df['rating'] = new_df['rating'].apply(remove_space)
new_df['listed_in'] = new_df['listed_in'].apply(separate)
new_df['description'] = new_df['description'].apply(remove_punc)

new_df.head()

type        director  \
title                                           
Dick Johnson Is Dead    movie  kirstenjohnson   
Blood & Water          tvshow                   
Ganglands              tvshow  julienleclercq   
Jailbirds New Orleans  tvshow                   
Kota Factory           tvshow                   

                                                                    cast  \
title                                                                      
Dick Johnson Is Dead                                                       
Blood & Water          amaqamata khosingema gailmabalane thabangmolab...   
Ganglands              samibouajila tracygotoas samueljouy nabihaakka...   
Jailbirds New Orleans                                                      
Kota Factory           mayurmore jitendrakumar ranjanraj alamkhan ahs...   

                      rating  \
title                          
Dick Johnson Is Dead   pg-13   
Blood & Water          tv-ma   
Ganglands              tv-ma   
Jailbirds New Orleans  tv-ma   
Kota Factory           tv-ma   

                                                               listed_in  \
title                                                                      
Dick Johnson Is Dead                                       documentaries   
Blood & Water                  internationaltvshows tvdramas tvmysteries   
Ganglands              crimetvshows internationaltvshows tvaction&adv...   
Jailbirds New Orleans                               docuseries realitytv   
Kota Factory             internationaltvshows romantictvshows tvcomedies   

                                                             description  
title                                                                     
Dick Johnson Is Dead   as her father nears the end of his life filmma...  
Blood & Water          after crossing paths at a party a cape town te...  
Ganglands              to protect his family from a powerful drug lor...  
Jailbirds New Orleans  feuds flirtations and toilet talk go down amon...  
Kota Factory           in a city of coaching centers known to train i...

In [10]:
new_df['bag_of_words'] = ''

# Combine all the words into 1 column
for i, row in enumerate(new_df.iterrows()):
    string = ''
    for col in new_df.columns:
        if row[1][col] == '':
            continue
        else:
            string += row[1][col] + ' '
            new_df['bag_of_words'][i] = string.strip()

new_df.drop(new_df.columns[:-1], axis=1, inplace=True)

In [11]:
new_df.head()

bag_of_words
title                                                                   
Dick Johnson Is Dead   movie kirstenjohnson pg-13 documentaries as he...
Blood & Water          tvshow amaqamata khosingema gailmabalane thaba...
Ganglands              tvshow julienleclercq samibouajila tracygotoas...
Jailbirds New Orleans  tvshow tv-ma docuseries realitytv feuds flirta...
Kota Factory           tvshow mayurmore jitendrakumar ranjanraj alamk...

In [12]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

<h1 style="font-family: Trebuchet MS; font-size: 25px; color: #a4161a; text-align: left; "><b>TF-IDF</b></h1>

**TF-IDF** stands for Term Frequency — Inverse Document Frequency. It tells the importance of a word. In a nutshell, The word that appear more frequently in the corpus, it will be considered less importance, hence the tfidf score will be lower. It goes the opposite way with less frequent word|

In [13]:
tfid = TfidfVectorizer()
tfid_matrix = tfid.fit_transform(new_df['bag_of_words'])

#tfid_matrix.vocabulary_

In [14]:
cosine_sim = cosine_similarity(tfid_matrix, tfid_matrix)
cosine_sim

array([[1.00000000e+00, 6.51109225e-03, 2.51508724e-02, ...,
        1.22541028e-02, 2.30160299e-02, 3.60339526e-02],
       [6.51109225e-03, 1.00000000e+00, 1.02129814e-02, ...,
        1.37342009e-03, 0.00000000e+00, 7.44741025e-04],
       [2.51508724e-02, 1.02129814e-02, 1.00000000e+00, ...,
        7.23719722e-03, 7.21477380e-03, 4.24499407e-02],
       ...,
       [1.22541028e-02, 1.37342009e-03, 7.23719722e-03, ...,
        1.00000000e+00, 1.99475996e-02, 5.30430706e-03],
       [2.30160299e-02, 0.00000000e+00, 7.21477380e-03, ...,
        1.99475996e-02, 1.00000000e+00, 4.54842501e-03],
       [3.60339526e-02, 7.44741025e-04, 4.24499407e-02, ...,
        5.30430706e-03, 4.54842501e-03, 1.00000000e+00]])

In [15]:
# Later on we will combine with similarity as a column
final_df = df[['title', 'type']]

In [16]:
def recommendation(title, total_result=5, threshold=0.5):
    # Get the index
    idx = final_df[final_df['title'] == title].index[0]
    # Create a new column for similarity, the value is different for each title you input
    final_df['similarity'] = cosine_sim[idx]
    sort_final_df = final_df.sort_values(by='similarity', ascending=False)[1:total_result+1]
    
    # You can set a threshold if you want to norrow the result down 
    #sort_final_df = sort_final_df[sort_final_df['similarity'] > threshold]
    
    # Is the title a movie or tv show?
    movies = sort_final_df['title'][sort_final_df['type'] == 'Movie']
    tv_shows = sort_final_df['title'][sort_final_df['type'] == 'TV Show']
    
    if len(movies) != 0:
        print('Similar Movie(s) list:')
        for i, movie in enumerate(movies):
            print('{}. {}'.format(i+1, movie))
        print()
    else:
        print('Similar Movie(s) list:')
        print('-\n')
        
    if len(tv_shows) != 0:
        print('Similar TV_show(s) list:')
        for i, tv_show in enumerate(tv_shows):
            print('{}. {}'.format(i+1, tv_show))
    else:
        print('Similar TV_show(s) list:')
        print('-')

<a id="4"></a>
<p style="background-color:#E50914;font-family:Trebuchet MS;font-weight:bold;color:#ffffff;font-size:40px;text-align:center;border-radius:100px 100px">Result</p>

In [17]:
recommendation('Breaking Bad')

Similar Movie(s) list:
1. The Show
2. The Book of Sun

Similar TV_show(s) list:
1. Better Call Saul
2. Marvel's The Punisher
3. Dare Me


In [18]:
recommendation('Narcos')

Similar Movie(s) list:
-

Similar TV_show(s) list:
1. Narcos: Mexico
2. Wild District
3. El Cartel
4. Miss Dynamite
5. Cocaine Cowboys: The Kings of Miami


In [19]:
recommendation('Chappie')

Similar Movie(s) list:
1. Real Steel
2. District 9
3. 2036 Origin Unknown
4. Singularity
5. AlphaGo

Similar TV_show(s) list:
-


In [20]:
recommendation('Stranger Things')

Similar Movie(s) list:
1. Safe Haven

Similar TV_show(s) list:
1. Beyond Stranger Things
2. The Umbrella Academy
3. Manifest
4. Prank Encounters


<h1 style="font-family: Trebuchet MS; font-size: 20px; color: #e85d04; text-align: center; "><b>If you find this notebook useful, give it a thumbs up 😉👍🏻</b></h1>
<h1 style="font-family: Trebuchet MS; font-size: 20px; color: #e85d04; text-align: center; "><b>Thank You</b></h1>

<h1 style="font-family: Trebuchet MS; font-size: 60px; color: #E50914; text-align: center; "><b>THE END</b></h1>